lib/sendgrid/mail.rb

```diff
@@ -28,34 +33,46 @@ def add_attachment(path, name = nil)
     end
 
     def to_h
-      payload = {
-        :from        => @from,
-        :fromname    => @from_name,
-        :subject     => @subject,
-        :to          => @to,
-        :toname      => @to_name,
-        :date        => @date,
-        :replyto     => @reply_to,
-        :cc          => @cc,
-        :bcc         => @bcc,
-        :text        => @text,
-        :html        => @html,
+      payload = {}
+
+      PAYLOAD_PARAMS.each do |payload_param|
+        payload[payload_mapping(payload_param)] = self.send(payload_param)
+      end
+
+      payload.merge!({
         :'x-smtpapi' => smtpapi_json,
         :files       => ({} unless @attachments.empty?)
-      }.reject {|k,v| v.nil?}
+      })
+
+      payload.reject! {|k,v| v.nil?}
+
+      assign_missing_to(payload)
+      extract_attachments(payload)
```

Comment:
Another benefit of this is that any key in the payload that has a nil value is not passed along. Since we don't know the entirety of the contract that sendgrid has under the hood, sending keys that have nil values is probably not advised.